Index: modules/pvm/src/main/java/org/jbpm/pvm/internal/job/JobImpl.java
===================================================================
--- modules/pvm/src/main/java/org/jbpm/pvm/internal/job/JobImpl.java	(revision 6337)
+++ modules/pvm/src/main/java/org/jbpm/pvm/internal/job/JobImpl.java	(revision 6338)
@@ -149,6 +149,7 @@
     return lockOwner;
   }
   /** @deprecated */
+  @Deprecated
   public Date getDueDate() {
     return getDuedate();
   }
Index: modules/pvm/src/main/java/org/jbpm/pvm/internal/xml/Parser.java
===================================================================
--- modules/pvm/src/main/java/org/jbpm/pvm/internal/xml/Parser.java	(revision 6337)
+++ modules/pvm/src/main/java/org/jbpm/pvm/internal/xml/Parser.java	(revision 6338)
@@ -272,6 +272,7 @@
   /** creates a new Parser with bindings and entities that can be maintained statically
    * in specialized subclasses of Parser.
    * @deprecated entities should be replaced by {@link #setSchemaResources(List)} */
+  @Deprecated
   public Parser(Bindings bindings, Map<String, Entity> entities) {
     initialize();
     this.bindings = bindings;
Index: modules/api/src/main/java/org/jbpm/api/job/Job.java
===================================================================
--- modules/api/src/main/java/org/jbpm/api/job/Job.java	(revision 6337)
+++ modules/api/src/main/java/org/jbpm/api/job/Job.java	(revision 6338)
@@ -40,6 +40,7 @@
   /** in case this is a timer, it is the time that the timer should fire, in case this 
    * is a message, it is null. 
    * @deprecated */
+  @Deprecated
   Date getDueDate();
 
   /** in case this is a timer, it is the time that the timer should fire, in case this 
Index: modules/examples/src/test/java/org/jbpm/examples/timer/businesstime/TimerBusinessTimeTest.java
===================================================================
--- modules/examples/src/test/java/org/jbpm/examples/timer/businesstime/TimerBusinessTimeTest.java	(revision 6337)
+++ modules/examples/src/test/java/org/jbpm/examples/timer/businesstime/TimerBusinessTimeTest.java	(revision 6338)
@@ -58,7 +58,7 @@
       .processInstanceId(processInstance.getId())
       .uniqueResult();
 
-    long difference = job.getDueDate().getTime() - now;
+    long difference = job.getDuedate().getTime() - now;
 
     // we do not know when this test will be run.  So the exact actual duedate of the timer
     // can not be calculated easily.  But we do know for sure that each working day only 
