Index: modules/jpdl/src/main/java/org/jbpm/jpdl/internal/xml/JpdlParser.java
===================================================================
--- modules/jpdl/src/main/java/org/jbpm/jpdl/internal/xml/JpdlParser.java	(revision 6380)
+++ modules/jpdl/src/main/java/org/jbpm/jpdl/internal/xml/JpdlParser.java	(revision 6381)
@@ -140,6 +140,10 @@
     Enumeration<URL> resourceUrls;
     try {
       resourceUrls = classLoader.getResources(resourceName);
+
+      if (!resourceUrls.hasMoreElements()) {
+        resourceUrls = JpdlParser.class.getClassLoader().getResources(resourceName);
+      }
     } catch (Exception e) {
       throw new JbpmException("couldn't get resource urls for "+resourceName, e);
     }
Index: modules/bpmn/src/main/java/org/jbpm/bpmn/parser/BpmnParser.java
===================================================================
--- modules/bpmn/src/main/java/org/jbpm/bpmn/parser/BpmnParser.java	(revision 6380)
+++ modules/bpmn/src/main/java/org/jbpm/bpmn/parser/BpmnParser.java	(revision 6381)
@@ -191,6 +191,10 @@
     Enumeration<URL> resourceUrls;
     try {
       resourceUrls = classLoader.getResources(resourceName);
+
+      if (!resourceUrls.hasMoreElements()) {
+        resourceUrls = BpmnParser.class.getClassLoader().getResources(resourceName);
+      }
     } catch (Exception e) {
       throw new JbpmException("couldn't get resource urls for "+resourceName, e);
     }
Index: modules/pvm/src/main/java/org/jbpm/pvm/internal/stream/ResourceStreamInput.java
===================================================================
--- modules/pvm/src/main/java/org/jbpm/pvm/internal/stream/ResourceStreamInput.java	(revision 6380)
+++ modules/pvm/src/main/java/org/jbpm/pvm/internal/stream/ResourceStreamInput.java	(revision 6381)
@@ -51,6 +51,11 @@
   public InputStream openStream() {
     ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
     InputStream stream = classLoader.getResourceAsStream(resource);
+
+    if (stream == null) {
+      stream = ResourceStreamInput.class.getClassLoader().getResourceAsStream(resource);
+    }
+
     if (stream==null) {
       throw new JbpmException("resource "+resource+" does not exist");
     }
Index: modules/pvm/src/main/java/org/jbpm/pvm/internal/wire/xml/WireParser.java
===================================================================
--- modules/pvm/src/main/java/org/jbpm/pvm/internal/wire/xml/WireParser.java	(revision 6380)
+++ modules/pvm/src/main/java/org/jbpm/pvm/internal/wire/xml/WireParser.java	(revision 6381)
@@ -276,6 +276,10 @@
       Enumeration<URL> resourceUrls;
       try {
         resourceUrls = classLoader.getResources(wireResource);
+
+        if (!resourceUrls.hasMoreElements()) {
+          resourceUrls = WireParser.class.getClassLoader().getResources(wireResource);
+        }
       } catch (Exception e) {
         throw new JbpmException("couldn't get resource urls for "+wireResource, e);
       }
