Index: modules/jpdl/src/test/java/org/jbpm/jpdl/parsing/ActivityParsingTest.java
===================================================================
--- modules/jpdl/src/test/java/org/jbpm/jpdl/parsing/ActivityParsingTest.java	(revision 6342)
+++ modules/jpdl/src/test/java/org/jbpm/jpdl/parsing/ActivityParsingTest.java	(revision 6343)
@@ -23,6 +23,9 @@
 
 import java.util.List;
 
+import org.jbpm.pvm.internal.client.ClientProcessDefinition;
+import org.jbpm.pvm.internal.model.ActivityImpl;
+import org.jbpm.pvm.internal.model.ProcessDefinitionImpl;
 import org.jbpm.pvm.internal.xml.Problem;
 
 
@@ -48,4 +51,20 @@
     );
     assertTextPresent("attribute <state name=\"\" is empty", problems.get(0).getMsg());
   }
+
+  public void testDescription() {
+    ClientProcessDefinition processDefinition = parse(
+      "<process name='p'>" +
+      "  <description>process definition description</description>" +
+      "  <start name='start'>" +
+      "    <description>start description</description>" +
+      "  </start>" +
+      "</process>"
+    );
+    assertEquals("process definition description", processDefinition.getDescription());
+
+    ProcessDefinitionImpl processDefinitionImpl = (ProcessDefinitionImpl) processDefinition;
+    ActivityImpl activity = processDefinitionImpl.getInitial();
+    assertEquals("start description", activity.getDescription());
+  }
 }
Index: modules/jpdl/src/main/java/org/jbpm/jpdl/internal/xml/JpdlParser.java
===================================================================
--- modules/jpdl/src/main/java/org/jbpm/jpdl/internal/xml/JpdlParser.java	(revision 6342)
+++ modules/jpdl/src/main/java/org/jbpm/jpdl/internal/xml/JpdlParser.java	(revision 6343)
@@ -282,7 +282,8 @@
       if ("on".equals(tagName) 
           || "timer".equals(tagName)
           || "swimlane".equals(tagName) 
-          || "migrate-instances".equals(tagName)) continue;
+          || "migrate-instances".equals(tagName)
+          || "description".equals(tagName)) continue;
 
       JpdlBinding activityBinding = (JpdlBinding) getBinding(nestedElement, CATEGORY_ACTIVITY);
       if (activityBinding == null) {
@@ -298,7 +299,7 @@
         parseTransitions(nestedElement, activity, parse);
         parseVariableDefinitions(nestedElement, parse, activity);
 
-        Element descriptionElement = XmlUtil.element(documentElement, "description");
+        Element descriptionElement = XmlUtil.element(nestedElement, "description");
         if (descriptionElement!=null) {
           String description = XmlUtil.getContentText(descriptionElement);
           activity.setDescription(description);
